Fix never returning gen_server:call(Server, sync) call

arcusfelis opened the pull request:
There are some records found in logs at startup:

```erlang
...there are more logs above...
when=2023-10-24T06:22:02.336073+00:00 level=error what=long_task_failed reason={{nodedown,'mongooseim@mongooseim-2.mongooseim.default.svc.cluster.local'},{gen_server,call,[<30886.526.0>,ping,infinity]}} pid=<0.2223.0> at=cets_long:run_tracked/2:63 stacktrace="gen_server:call/3:401 cets_long:run_tracked/2:54 lists:foreach_1/2:1686 cets:'-handle_call/3-fun-1-'/2:441 proc_lib:init_p/3:226" server=<30886.526.0> msg=ping pid=<30886.526.0> node=mongooseim@mongooseim-2.mongooseim.default.svc.cluster.local
...there are more logs here too...
when=2023-10-24T06:22:12.340256+00:00 level=warning what=long_task_progress pid=<0.2222.0> at=cets_long:monitor_loop/5:88 time_ms=10004 server=cets_session msg=sync pid=<0.526.0> node=mongooseim@mongooseim-0.mongooseim.default.svc.cluster.local current_stacktrace="[{gen,do_call,4,[{file,\"gen.erl\"},{line,240}]},{gen_server,call,3,[{file,\"gen_server.erl\"},{line,397}]},{cets_long,run_tracked,2,[{file,\"/home/circleci/project/_build/default/lib/cets/src/cets_long.erl\"},{line,54}]},{ejabberd_sm_cets,cleanup,1,[{file,\"/home/circleci/project/src/ejabberd_sm_cets.erl\"},{line,79}]},{timer,tc,3,[{file,\"timer.erl\"},{line,295}]},{ejabberd_sm,handle_call,3,[{file,\"/home/circleci/project/src/ejabberd_sm.erl\"},{line,474}]},{gen_server,try_handle_call,4,[{file,\"gen_server.erl\"},{line,1113}]},{gen_server,handle_msg,6,[{file,\"gen_server.erl\"},{line,1142}]}]"
when=2023-10-24T06:22:17.342470+00:00 level=warning what=long_task_progress pid=<0.2222.0> at=cets_long:monitor_loop/5:88 time_ms=15006 server=cets_session msg=sync pid=<0.526.0> node=mongooseim@mongooseim-0.mongooseim.default.svc.cluster.local current_stacktrace="[{gen,do_call,4,[{file,\"gen.erl\"},{line,240}]},{gen_server,call,3,[{file,\"gen_server.erl\"},{line,397}]},{cets_long,run_tracked,2,[{file,\"/home/circleci/project/_build/default/lib/cets/src/cets_long.erl\"},{line,54}]},{ejabberd_sm_cets,cleanup,1,[{file,\"/home/circleci/project/src/ejabberd_sm_cets.erl\"},{line,79}]},{timer,tc,3,[{file,\"timer.erl\"},{line,295}]},{ejabberd_sm,handle_call,3,[{file,\"/home/circleci/project/src/ejabberd_sm.erl\"},{line,474}]},{gen_server,try_handle_call,4,[{file,\"gen_server.erl\"},{line,1113}]},{gen_server,handle_msg,6,[{file,\"gen_server.erl\"},{line,1142}]}]"
```

Generally, that nodedown is caused by `global` module doing that:

```erlang
when=2023-10-24T06:22:10.304831+00:00 level=warning pid=<0.59.0> at=: unstructured_log="'global' at node 'mongooseim@mongooseim-0.mongooseim.default.svc.cluster.local' requested disconnect from node 'mongooseim@mongooseim-2.mongooseim.default.svc.cluster.local' in order to prevent overlapping partitions"
when=2023-10-24T06:22:10.306128+00:00 level=warning pid=<0.59.0> at=: unstructured_log="'global' at node 'mongooseim@mongooseim-0.mongooseim.default.svc.cluster.local' requested disconnect from node 'mongooseim@mongooseim-1.mongooseim.default.svc.cluster.local' in order to prevent overlapping partitions"
```

But in case global disconnects one of the pinged nodes, `sync` call could crash and never return to the client.

The issue is related to "MIM-2073 Multiple errors in the logs on cluster startup with CETS", but does not fully fixes all the possible errors in MIM-2073. 

So, this PR has a very limited scope:
- Fix never returning sync call
- Ensure that we send the reply to the caller of the sync call 
- Add cets_sync_returns_when_ping_crashes testcase

It does not improve logs or fixes global behaviour - there would be separate PRs for that.

arcusfelis committed:
Fix never returning sync call

Ensure that we send the reply to the caller of the sync call
Add cets_sync_returns_when_ping_crashes testcase

chrzaszcz commented on src/cets.erl:
```diff
@@ -438,7 +438,9 @@ handle_call(get_leader, _From, State = #{leader := Leader}) ->
-        lists:foreach(fun ping/1, Servers),
+        %% If ping crashes, the caller would not receive a reply.
+        %% So, we have to use catch to still able to reply with ok.
+        lists:foreach(fun(Server) -> catch ping(Server) end, Servers),
```
So we decide that we still want to return `ok` even though `ping` crashed, which means we might be not fully synced? I want to make sure what the intention is.

arcusfelis replied:
`sync` is used to ensure that all the servers in the cluster consumed their mail box.
But we only care about messages that were sent before the client code executed `cets:sync/1` (in our case, the "client" code is `cets_join`).

That's actually it, we do not give more guarantees. `ping` here is just a sync gen_server call.
If the connection fails or one of the servers crashes - the code that called sync, could continue execution.
Usually that means that later the local gen_server would receive `DOWN` from the remote gen_server, and could start the cleaning procedure.

Actually, you are right, we could just cancel joining earlier, if any of the servers gets lost. So, we would not need to get the data dump in this case - a bit of optimisation here too.

We could return `{error, {bad_servers, List}}` from sync and of there is an error, `cets_join` would exit earlier (and it would be retried later again).

arcusfelis committed:
Rename cets:sync to cets:ping_all

In case ping_all fails to reach one of the servers - return an error.
In this case cets_join would return an error and stop joining earlier
(to avoid making a table dump first)

arcusfelis committed:
Unload meck to avoid already_started error in tests

arcusfelis committed:
Run cets_ping_all_returns_when_ping_crashes testcase without logger too

To improve code coverage

chrzaszcz reviewed (approved):
Looks ok :ok_hand:

arcusfelis committed:
Use task_failed format for errors in join_interrupted_when_ping_crashes
